src/vs/workbench/contrib/notebook/browser/notebookExecutionStateServiceImpl.ts

```diff
@@ -142,14 +143,87 @@ export class NotebookExecutionStateService extends Disposable implements INotebo
-		this._lastFailedCells.set(notebook, cellHandle);
-		this._onDidChangeLastRunFailState.fire({ failed: true, notebook });
+	private _setLastFailedCell(notebookURI: URI, cellHandle: number) {
+		const prevLastFailedCellInfo = this._lastFailedCells.get(notebookURI);
+		const newLastFailedCellInfo: IFailedCellInfo = { cellHandle: cellHandle, visible: true };
+		if (prevLastFailedCellInfo) {
+			newLastFailedCellInfo.disposable = prevLastFailedCellInfo.disposable;
+		} else {
+			newLastFailedCellInfo.disposable = this._getFailedCellListener(notebookURI);
+		}
+		this._lastFailedCells.set(notebookURI, newLastFailedCellInfo);
+
+		this._onDidChangeLastRunFailState.fire({ visible: true, notebook: notebookURI });
+	}
+
+	private _setLastFailedCellVisibility(notebookURI: URI, visible: boolean) {
+		const lastFailedCellInfo = this._lastFailedCells.get(notebookURI);
+
+		if (lastFailedCellInfo) {
+			this._lastFailedCells.set(notebookURI, {
+				cellHandle: lastFailedCellInfo.cellHandle,
+				disposable: lastFailedCellInfo.disposable,
+				visible: visible,
+			});
+		}
+
+		this._onDidChangeLastRunFailState.fire({ visible: visible, notebook: notebookURI });
+	}
+
+	private _clearLastFailedCell(notebookURI: URI) {
+
+		const lastFailedCellInfo = this._lastFailedCells.get(notebookURI);
+
+		if (lastFailedCellInfo) {
+			lastFailedCellInfo.disposable?.dispose();
+			this._lastFailedCells.delete(notebookURI);
+		}
+
+		this._onDidChangeLastRunFailState.fire({ visible: false, notebook: notebookURI });
 	}
 
-	private _clearLastFailedCell(notebook: URI) {
-		this._lastFailedCells.delete(notebook);
-		this._onDidChangeLastRunFailState.fire({ failed: false, notebook: notebook });
+	private _getFailedCellListener(notebookURI: URI) {
+		const notebook = this._notebookService.getNotebookTextModel(notebookURI);
+		let lastFailedCellDisposableListener;
+		if (notebook) {
+			lastFailedCellDisposableListener = combinedDisposable(
+				notebook.onWillAddRemoveCells((e: NotebookTextModelWillAddRemoveEvent) => {
+					const lastFailedCell = this._lastFailedCells.get(notebook.uri)?.cellHandle;
+					if (lastFailedCell !== undefined) {
+						const lastFailedCellPos = notebook.cells.findIndex(c => c.handle === lastFailedCell);
+						e.rawEvent.changes.some(([start, deleteCount]) => {
+							if (deleteCount) {
+								if (lastFailedCellPos >= start && lastFailedCellPos < start + deleteCount) {
+									this._setLastFailedCellVisibility(notebook.uri, false);
+									return true;
+								}
+							}
+							return false;
+						});
+					}
+				}),
+				notebook.onDidChangeContent((events) => {
+					const lastFailedCell = this._lastFailedCells.get(notebook.uri)?.cellHandle;
+					if (lastFailedCell !== undefined) {
+						events.rawEvents.forEach((e) => {
+							if (e.kind === NotebookCellsChangeType.ModelChange) {
+								e.changes.some(([start, deleteCount]) => {
+									if (deleteCount === 0) {
+										const addHandle = notebook.cells[start].handle;
```

Comment:
This is only checking one item, when multiple could be inserted with one event. Also, items could be added and removed at the same time